libethcore/EthashAux.cpp

```diff
@@ -32,26 +32,26 @@ EthashAux& EthashAux::get()
 int EthashAux::toEpoch(h256 const& _seedHash)
 {
     EthashAux& ethash = EthashAux::get();
-    Guard l(ethash.x_epochs);
-    int epoch = 0;
-    auto epochIter = ethash.m_epochs.find(_seedHash);
-    if (epochIter == ethash.m_epochs.end())
+
+    if (_seedHash != ethash.m_cached_seed)
     {
-        for (h256 h; h != _seedHash && epoch < 2048;
-             ++epoch, h = sha3(h), ethash.m_epochs[h] = epoch)
+        // Find epoch number corresponding to given seed hash.
+        int epoch = 0;
+        for (h256 h; h != _seedHash && epoch < 2048; ++epoch, h = sha3(h))
```

Comment:
Since epoch will always go up, why not start the loop with cached seed at cached epoch?

Reply:
Maybe only start at 0 if epoch is less than cached epoch

Reply:
toEpoch will only be called by single stratum thread, toSeed will be called asynchronously by multiple miner threads by using eval.

Reply:
However, epoch is an simple integer variable easily dealt with using atomic if you want to make toEpoch thread safe without a lock.

Reply:
I implemented better version in https://github.com/chfast/ethash/pull/13. Will try to use it here later on.

It takes < 50 ms to find all epoch number up to 30000 for sequential search.